### PR TITLE
Add support for timeuuid v7

### LINKED ISSUE #11055: Suuport UUID v7 as TimeUUID
[UUID v7](https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#name-uuid-version-7) implementations are being rolled out. Projects are starting to use it instead of UUID v1. At the moment, ScyllaDB does not support UUIDv7 as `timeuuid` and can therefore not enjoy its indexing benefits.

The code base is quite complex. But the error message `Unsupported UUID version` got me to this line:
https://github.com/scylladb/scylla/blob/d3c6e72c6934cb7b8f16231ea2fa7075d855202f/types.cc#L1650

Not sure if only this line should be modified to support v7?

### PR DESCRIPTION
**Pull request by bhalevy**
Based on https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format-04

UUID Version 7 as defined in the above Internet Draft provides benefits over the exiting time uuid format (v1).
It is easier and faster to construct and and compare as it begins (in memory order) with the unix timestamp in milleseconds encoded using 48 bits, unlike uuid v1 which requires scrambling of the timestamp part of the uuid.

In addition, the string representation of uuid v7 also sorts lexicographically for uuids with unique timestamps (in milliseconds, and having the same sign) making them more suitable to be used as unique sstable generations.

The downside of using time uuid v7 is that its sub-milliseconds resolution is inferior to v1.  v7 has only 12 bits set for the sub-milliseconds part, which we use in this series to store a unique sequence number for repeated calls to get_time_UUID_v7, while v1 has 14 bits that are used in theory to count decimicroseconds (100 ns) units, yet we use them for uniqueness in a similar way and base even the v1 time uuids on the system clock in milliseconds resolution.

This series allows users to store and retrieve timeuuid values in uuid v7 format.
These can be compared in an interchangeable way with time uuid v1 values.

v1 <=> v1 comparisons retain the existing comparison algorithm.
v7 <=> v7 comparisons use the optimized, byte-level unsigned compare.
v1 <=> v7 comparisons compute a normalized timestamp from the two values and compare them as signed int64_t offsets from the unix epoch.

Fixes #11055

### CODE REVIEW

**Comment by bhalevy** [comment]
Cc @tzach

**Comment by bhalevy** [comment]
In v2 ([f8e0ec7](https://github.com/scylladb/scylladb/commit/f8e0ec7c59f7cc0378db92a3febb0cb41028553f)):
- fix and optimize v7 comparison (against itself and against v1)

**Comment by avikivity** [comment]
Since the draft is not yet approved (AFAICT), should we implement it? It may change or be rejected.

**Comment by bhalevy** [comment]
> Since the draft is not yet approved (AFAICT), should we implement it? It may change or be rejected.

IETF drafts may take many years to proceed and become RFCs.
One of the requirements for that to happen is proving that there is community interest and demonstrating interoperability between at least 2 independent implementations, so if nobody implements it until it becomes a standard it will never become a standard.

Last change to the v7 format was in March this year (https://www.ietf.org/rfcdiff?url2=draft-peabody-dispatch-new-uuid-format-03.txt) and the following version (https://www.ietf.org/rfcdiff?url2=draft-peabody-dispatch-new-uuid-format-04.txt) had no change to the v7 format.

Bottom line is that I think it would help the overall process if we provide an experimental implementation of UUID version 7 and provide feedback to the IETF working group if needed based on our experience.  Then we can make it GA once the draft advances to RFC status.

**Comment by avikivity** [comment]
> > Since the draft is not yet approved (AFAICT), should we implement it? It may change or be rejected.
> 
> IETF drafts may take many years to proceed and become RFCs. One of the requirements for that to happen is proving that there is community interest and demonstrating interoperability between at least 2 independent implementations, so if nobody implements it until it becomes a standard it will never become a standard.
> 
> Last change to the v7 format was in March this year (https://www.ietf.org/rfcdiff?url2=draft-peabody-dispatch-new-uuid-format-03.txt) and the following version (https://www.ietf.org/rfcdiff?url2=draft-peabody-dispatch-new-uuid-format-04.txt) had no change to the v7 format.
> 
> Bottom line is that I think it would help the overall process if we provide an experimental implementation of UUID version 7 and provide feedback to the IETF working group if needed based on our experience. Then we can make it GA once the draft advances to RFC status.

Experimental implementation maybe, but if we start using it and embedding it in persistent data, and then the format changes, what?

**Comment by tzach** [comment]
@bhalevy please update https://github.com/scylladb/scylladb/blob/master/docs/cql/types.rst to match this update

In particular:
"Version 1 UUID, generally used as a “conflict-free” timestamp. See Working with UUIDs for details"

and 
https://docs.scylladb.com/stable/cql/types.html#working-with-uuids

**Comment by bhalevy** [comment]
In [e5e79a3](https://github.com/scylladb/scylladb/commit/e5e79a3a0ecf2036a9d3d3d262642e6c9ab69562):
- rebased
- added "docs: cql/types: timeuuid: allow also Version 7 UUID"

**Comment by avikivity** [comment]
I still don't think we should apply it before the draft is accepted (unfortunately "RFC" can mean an accepted change too).

**Comment by bhalevy** [comment]
okay, I"ll track the draft's progress

**Comment by mykaul** [comment]
> I still don't think we should apply it before the draft is accepted (unfortunately "RFC" can mean an accepted change too).

It did not stop Cassandra for getting this in (https://issues.apache.org/jira/browse/CASSANDRA-17048 seems to be fixed for 4.1). I'm unsure how much activity there is on the RFC and if it's still in a heated debate or should be ratified or approved soon or whatnot, of course.

**Comment by bhalevy** [comment]
The draft is managed on github: https://github.com/ietf-wg-uuidrev/rfc4122bis

Discussion thread: https://mailarchive.ietf.org/arch/browse/uuidrev/?q=draft-ietf-uuidrev-rfc4122bis

Latest update:
```
Re: [Uuidrev] Post-IETF 115 and Draft 01 Reviews
"Kyzer Davis (kydavis)" <kydavis@cisco.com> Tue, 13 December 2022 18:51 UTC[Show header](https://mailarchive.ietf.org/arch/browse/uuidrev/?q=draft-ietf-uuidrev-rfc4122bis#)

Quick Update,  <mailto:brad@peabody.io> @Brad Peabody knocked out two more
PRs we can get into Draft 01.

*	Some UUIDv6 text update:

*	https://github.com/ietf-wg-uuidrev/rfc4122bis/pull/34

*	URN Template Update:

*	https://github.com/ietf-wg-uuidrev/rfc4122bis/pull/35

 

These go along with my open PR for the other draft items for Draft 01

*	https://github.com/ietf-wg-uuidrev/rfc4122bis/pull/32



I will leave these in PR state until we get the go ahead on the next interim
to merge. In the meantime, the link below can be used to see all items Brad
and I have slated for Consensus Request:
https://github.com/ietf-wg-uuidrev/rfc4122bis/issues?q=is%3Aissue+is%3Aopen+
label%3A%22Consensus+Requested%22

 

Outstanding items still in need of discussion or feedback:

*	Draft 01: Security Considerations for Running out of Random #26
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/26> 
*	Reserve variant for Omni-UUID (Max-UUID) #16
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/16> 

 

Thanks,

 

From: Uuidrev [<uuidrev-bounces@ietf.org>](mailto:&lt;uuidrev-bounces@ietf.org&gt;) On Behalf Of Kyzer Davis (kydavis)
Sent: Monday, November 28, 2022 9:19 AM
To: Kyzer Davis (kydavis) [<kydavis=40cisco.com@dmarc.ietf.org>](mailto:&lt;kydavis=40cisco.com@dmarc.ietf.org&gt;);
[uuidrev@ietf.org](mailto:uuidrev@ietf.org)
Subject: Re: [Uuidrev] Post-IETF 115 and Draft 01 Reviews

 

Second poke to the mailer.

 

Any objections to anything in the Draft 01 PR?
https://github.com/ietf-wg-uuidrev/rfc4122bis/pull/32
This is probably all I will get done for this draft and bump anything
leftover to Draft 02.

 

Specific Items being fixed in Draft 01:

*	Draft 01: Mixed Case Spelling error
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/18>  #18
*	Draft 01: Add
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/19> "UUIDs that Do Not
Identify the Host as well" reference to security considerations #19
*	Draft 01: Out of Place Distributed node text
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/20>  #20
*	Draft 01: Figure 2 and 3 Fix Title
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/22>  #22
*	Draft 01: Verify ABNF formatting against RFC5234
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/24>  #24
*	Draft 01: Bump ABNF reference to RFC 5234
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/25>  #25
*	Draft 01: Modify v8 SHOULD NOT to MUST NOT
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/27>  #27
*	Draft 01: Remove
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/29> "time-based"
constraint from version 8 UUID #29
*	Draft 01: Further clarify v7 field description #125
<https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/30>  #30

 

Thanks,

 

From: Uuidrev [<uuidrev-bounces@ietf.org](mailto:&lt;uuidrev-bounces@ietf.org) <mailto:uuidrev-bounces@ietf.org> >
On Behalf Of Kyzer Davis (kydavis)
Sent: Monday, November 14, 2022 5:23 PM
To: [uuidrev@ietf.org](mailto:uuidrev@ietf.org) <mailto:uuidrev@ietf.org> 
Subject: [Uuidrev] Post-IETF 115 and Draft 01 Reviews

 

Hello all,

 

As agreed during IETF 115:

I will be creating pull requests to review all ongoing changes before
merging into draft 01.


I went ahead and created issues on the tracker for all of the Draft 01 items
and tagged them all as "Draft 01" in the title so filtering is easy.

*
https://github.com/ietf-wg-uuidrev/rfc4122bis/issues?q=is%3Aissue+is%3Aopen+
Draft+01

 

I knocked out the bulk of the smaller items to practice the process and PR
is open here for review:
https://github.com/ietf-wg-uuidrev/rfc4122bis/pull/32

*	Each issue/change is it's own commit so we can look at just the
modifications in that context. Example:
https://github.com/ietf-wg-uuidrev/rfc4122bis/commit/a4673042038672372c9c350
176ce740fc8cfdbcd
*	I labeled all of these as "Consensus Requested" so I can track my
edit todo.

 

Further I also labeled all Errata as either Technical or Grammar so you may
use the link below when reviewing the draft to verify, we squashed the
errata in the new document.

*	Technical Errata:
https://github.com/ietf-wg-uuidrev/rfc4122bis/issues?q=is%3Aissue+is%3Aopen+
label%3ATechnical
*	Grammatical Errata:
https://github.com/ietf-wg-uuidrev/rfc4122bis/issues?q=is%3Aissue+is%3Aopen+
label%3ATechnical

 

Thanks,

 

---

Kyzer Davis
```

**Comment by AugustDev** [comment]
Any progress on this? Seems like a powerful feature.

**Comment by kcking** [comment]
FWIW it looks like the uuidv7 draft has been [approved ](https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7)

**Comment by bhalevy** [comment]
Replaced with https://github.com/scylladb/scylladb/pull/19642
